pkg/controller/direct/monitoring/maputils.go

```diff
@@ -0,0 +1,179 @@
+package monitoring
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/durationpb"
+)
+
+type MapContext struct {
+	errs []error
+}
+
+func (c *MapContext) Errorf(msg string, args ...interface{}) {
+	c.errs = append(c.errs, fmt.Errorf(msg, args...))
+}
+
+func (c *MapContext) Err() error {
+	return errors.Join(c.errs...)
+}
+
+func Slice_ToProto[T, U any](ctx *MapContext, in []T, mapper func(ctx *MapContext, in *T) *U) []*U {
+	if in == nil {
+		return nil
+	}
+
+	outSlice := make([]*U, 0, len(in))
+	for _, inItem := range in {
+		outItem := mapper(ctx, &inItem)
+		outSlice = append(outSlice, outItem)
+	}
+	return outSlice
+}
+
+func Slice_FromProto[T, U any](ctx *MapContext, in []*T, mapper func(ctx *MapContext, in *T) *U) []U {
+	if in == nil {
+		return nil
+	}
+
+	outSlice := make([]U, 0, len(in))
+	for _, inItem := range in {
+		outItem := mapper(ctx, inItem)
+		outSlice = append(outSlice, *outItem)
+	}
+	return outSlice
+}
+
+type ProtoEnum interface {
+	~int32
+	Descriptor() protoreflect.EnumDescriptor
+}
+
+func Enum_ToProto[U ProtoEnum](ctx *MapContext, in *string) U {
+	var defaultU U
+	descriptor := defaultU.Descriptor()
+
+	inValue := ValueOf(in)
+	if inValue == "" {
+		unspecifiedValue := U(int32(0)) // defaultU.New(protoreflect.EnumNumber(0)).(U)
```

Comment:
for my own learning, here and everywhere else, why not int64?

Reply:
Hmmm, good question.  It shouldn't matter in go, because constant values are untyped (or something).  Removed the int32 entirely, and it seems to still work :-)

Reply:
(proto.EnumNumber is int32, so it is actually int32 at the proto level, though proto int64 vs int32 is largely academic anyway as they both use variable length encoding, I believe)